src/service-worker/core.js

```diff
@@ -225,6 +244,109 @@ export function generateFallbackClientId(referrer) {
 }
 
 /**
+ * Fetches a URL, and stores it into the cache if the response is valid.
+ * Repeated fetches of the same URL will be batched into a single request while
+ * the first is still fetching.
+ *
+ * @param {!Cache} cache
+ * @param {!Request} request
+ * @return {!Promise<!Response>}
+ * @visibleForTesting
+ */
+export function fetchAndCache(cache, request) {
+  const url = request.url;
+
+  // Batch fetches. Mainly for the /diversions endpoint.
+  if (fetchPromises[url]) {
+    return fetchPromises[url];
+  }
+
+  return fetchPromises[url] = cache.match(request)
+      .then(response => {
+        if (response && !expired(response)) {
+          delete fetchPromises[url];
+          return response;
+        }
+
+        return fetch(request).then(response => {
+          delete fetchPromises[url];
+
+          // Did we receive a invalid response?
+          if (!response.ok) {
+            throw new Error(`failed fetching ${url}`);
+          }
+
+          // You must clone to prevent double reading the body.
+          cache.put(request, response.clone());
+          return response;
+        }, err => {
+          delete fetchPromises[url];
+          throw err;
+        });
+      });
+}
+
+/**
+ * Checks if a (valid) response has expired.
+ *
+ * @param {!Response} response
+ * @return {boolean}
+ * @visibleForTesting
+ */
+export function expired(response) {
+  const {headers} = response;
+  let date;
+  let age = 0;
+
+  if (headers.has('date') && headers.has('cache-control')) {
+    const maxAge = /max-age=(\d+)/i.exec(headers.get('cache-control'));
+    date = headers.get('date');
+    age = maxAge ? maxAge[1] * 1000 : -Infinity;
+  } else {
+    date = headers.get('expires');
+  }
+
+  return Date.now() >= Number(new Date(date)) + age;
```

Comment:
> I'd consider adding a shortcut for "reasonably close to infinity".

What do you mean?

> Wouldn't this consider everything expired if the machine's clock is in the future?

Hmm, I can override the header on the `Date` header on our side, ensuring the response uses the user's clock time.